In [1]:
import pandas as pd
from utils import get_Xy, preprocess, model_generator, evaluate

In [2]:
%load_ext autoreload
%autoreload 2

In [3]:
df = pd.read_csv('../data/predictive_maintenance.csv')

In [4]:
X_train, X_test, y_train, y_test = get_Xy(df, 'Failure Type', binary=['No Failure'], drop_cols=['UDI', 'Product ID'])

In [5]:
preprocessor = preprocess(X_train, y_train, label_cols=['Target'])

In [6]:
pipeline = model_generator(X_train, y_train, preprocessor, models=None)

Fitting 5 folds for each of 80 candidates, totalling 400 fits
[CV 1/5] END leaf_size=10, n_neighbors=1, weights=uniform;, score=0.997 total time=   0.9s
[CV 2/5] END leaf_size=10, n_neighbors=1, weights=uniform;, score=1.000 total time=   0.6s
[CV 3/5] END leaf_size=10, n_neighbors=1, weights=uniform;, score=0.999 total time=   0.6s
[CV 4/5] END leaf_size=10, n_neighbors=1, weights=uniform;, score=0.998 total time=   0.6s
[CV 5/5] END leaf_size=10, n_neighbors=1, weights=uniform;, score=0.998 total time=   0.6s
[CV 1/5] END leaf_size=10, n_neighbors=1, weights=distance;, score=0.997 total time=   0.7s
[CV 2/5] END leaf_size=10, n_neighbors=1, weights=distance;, score=1.000 total time=   0.5s
[CV 3/5] END leaf_size=10, n_neighbors=1, weights=distance;, score=0.999 total time=   0.6s
[CV 4/5] END leaf_size=10, n_neighbors=1, weights=distance;, score=0.998 total time=   0.6s
[CV 5/5] END leaf_size=10, n_neighbors=1, weights=distance;, score=0.998 total time=   0.6s
[CV 1/5] END leaf_size=

/Users/adellehousker/anaconda3/envs/dev/lib/python3.10/site-packages/sklearn/ensemble/_weight_boosting.py:519: FutureWarning: The SAMME.R algorithm (the default) is deprecated and will be removed in 1.6. Use the SAMME algorithm to circumvent this warning.
  warnings.warn(


Fitting 5 folds for each of 4 candidates, totalling 20 fits
[CV 1/5] END .....................kernel=linear;, score=0.999 total time=   4.7s
[CV 2/5] END .....................kernel=linear;, score=0.999 total time=   4.9s
[CV 3/5] END .....................kernel=linear;, score=1.000 total time=   4.8s
[CV 4/5] END .....................kernel=linear;, score=0.998 total time=   4.4s
[CV 5/5] END .....................kernel=linear;, score=0.999 total time=   4.6s
[CV 1/5] END .......................kernel=poly;, score=0.999 total time=  14.8s
[CV 2/5] END .......................kernel=poly;, score=1.000 total time=  14.8s
[CV 3/5] END .......................kernel=poly;, score=1.000 total time=  14.6s
[CV 4/5] END .......................kernel=poly;, score=0.999 total time=  14.6s
[CV 5/5] END .......................kernel=poly;, score=1.000 total time=  14.5s
[CV 1/5] END ........................kernel=rbf;, score=0.999 total time=  11.2s
[CV 2/5] END ........................kernel=rbf;,

In [7]:
evals = evaluate(pipeline, X_test, y_test)

In [8]:
for e in evals:
    print(e['model'], '-' * (54 - len(e['model'])))
    print(e['report'])

Logistic Regression -----------------------------------
              precision    recall  f1-score   support

    negative       0.95      0.97      0.96        71
    positive       1.00      1.00      1.00      1929

    accuracy                           1.00      2000
   macro avg       0.97      0.98      0.98      2000
weighted avg       1.00      1.00      1.00      2000

KNN ---------------------------------------------------
              precision    recall  f1-score   support

    negative       0.89      0.92      0.90        71
    positive       1.00      1.00      1.00      1929

    accuracy                           0.99      2000
   macro avg       0.94      0.96      0.95      2000
weighted avg       0.99      0.99      0.99      2000

Decision Tree -----------------------------------------
              precision    recall  f1-score   support

    negative       0.92      0.93      0.92        71
    positive       1.00      1.00      1.00      1929

    accuracy  

--- WITHOUT OVERSAMPLING ---

Logistic Regression -----------------------------------
              precision    recall  f1-score   support

    negative       0.95      0.96      0.95        76
    positive       1.00      1.00      1.00      1924

    accuracy                           1.00      2000
   macro avg       0.97      0.98      0.98      2000
weighted avg       1.00      1.00      1.00      2000

KNN ---------------------------------------------------
              precision    recall  f1-score   support

    negative       0.96      0.89      0.93        76
    positive       1.00      1.00      1.00      1924

    accuracy                           0.99      2000
   macro avg       0.98      0.95      0.96      2000
weighted avg       0.99      0.99      0.99      2000

Decision Tree -----------------------------------------
              precision    recall  f1-score   support

    negative       0.93      0.93      0.93        76
    positive       1.00      1.00      1.00      1924

    accuracy                           0.99      2000
   macro avg       0.97      0.97      0.97      2000
weighted avg       0.99      0.99      0.99      2000

Random Forest -----------------------------------------
              precision    recall  f1-score   support

    negative       0.95      0.96      0.95        76
    positive       1.00      1.00      1.00      1924

    accuracy                           1.00      2000
   macro avg       0.97      0.98      0.98      2000
weighted avg       1.00      1.00      1.00      2000

Extremely Random Trees --------------------------------
              precision    recall  f1-score   support

    negative       0.95      0.96      0.95        76
    positive       1.00      1.00      1.00      1924

    accuracy                           1.00      2000
   macro avg       0.97      0.98      0.98      2000
weighted avg       1.00      1.00      1.00      2000

Gradient Boosting -------------------------------------
              precision    recall  f1-score   support

    negative       0.95      0.96      0.95        76
    positive       1.00      1.00      1.00      1924

    accuracy                           1.00      2000
   macro avg       0.97      0.98      0.98      2000
weighted avg       1.00      1.00      1.00      2000

AdaBoost ----------------------------------------------
              precision    recall  f1-score   support

    negative       0.95      0.95      0.95        76
    positive       1.00      1.00      1.00      1924

    accuracy                           1.00      2000
   macro avg       0.97      0.97      0.97      2000
weighted avg       1.00      1.00      1.00      2000

SVM ---------------------------------------------------
              precision    recall  f1-score   support

    negative       0.95      0.96      0.95        76
    positive       1.00      1.00      1.00      1924

    accuracy                           1.00      2000
   macro avg       0.97      0.98      0.98      2000
weighted avg       1.00      1.00      1.00      2000



--- WITH OVERSAMPLING ---

Logistic Regression -----------------------------------
              precision    recall  f1-score   support

    negative       0.95      0.97      0.96        71
    positive       1.00      1.00      1.00      1929

    accuracy                           1.00      2000
   macro avg       0.97      0.98      0.98      2000
weighted avg       1.00      1.00      1.00      2000

KNN ---------------------------------------------------
              precision    recall  f1-score   support

    negative       0.89      0.92      0.90        71
    positive       1.00      1.00      1.00      1929

    accuracy                           0.99      2000
   macro avg       0.94      0.96      0.95      2000
weighted avg       0.99      0.99      0.99      2000

Decision Tree -----------------------------------------
              precision    recall  f1-score   support

    negative       0.92      0.93      0.92        71
    positive       1.00      1.00      1.00      1929

    accuracy                           0.99      2000
   macro avg       0.96      0.96      0.96      2000
weighted avg       0.99      0.99      0.99      2000

Random Forest -----------------------------------------
              precision    recall  f1-score   support

    negative       0.96      0.97      0.97        71
    positive       1.00      1.00      1.00      1929

    accuracy                           1.00      2000
   macro avg       0.98      0.99      0.98      2000
weighted avg       1.00      1.00      1.00      2000

Extremely Random Trees --------------------------------
              precision    recall  f1-score   support

    negative       0.96      0.97      0.97        71
    positive       1.00      1.00      1.00      1929

    accuracy                           1.00      2000
   macro avg       0.98      0.99      0.98      2000
weighted avg       1.00      1.00      1.00      2000

Gradient Boosting -------------------------------------
              precision    recall  f1-score   support

    negative       0.92      0.97      0.95        71
    positive       1.00      1.00      1.00      1929

    accuracy                           1.00      2000
   macro avg       0.96      0.98      0.97      2000
weighted avg       1.00      1.00      1.00      2000

AdaBoost ----------------------------------------------
              precision    recall  f1-score   support

    negative       0.87      0.97      0.92        71
    positive       1.00      0.99      1.00      1929

    accuracy                           0.99      2000
   macro avg       0.94      0.98      0.96      2000
weighted avg       0.99      0.99      0.99      2000

SVM ---------------------------------------------------
              precision    recall  f1-score   support

    negative       0.96      0.97      0.97        71
    positive       1.00      1.00      1.00      1929

    accuracy                           1.00      2000
   macro avg       0.98      0.99      0.98      2000
weighted avg       1.00      1.00      1.00      2000

Naive Bayes -------------------------------------------
              precision    recall  f1-score   support

    negative       0.09      0.18      0.12        71
    positive       0.97      0.93      0.95      1929

    accuracy                           0.90      2000
   macro avg       0.53      0.56      0.53      2000
weighted avg       0.94      0.90      0.92      2000

